File: packages/dsh-stream-state/swift/LoadState.swift
//
//  LoadState.swift
//  Swift mirror - the shell's explicit load state machine.
//
//  WHY THIS FILE EXISTS. Three independent booleans cannot express the load state
//  safely: a `didStartLoading` latch with no reset point, a `webViewContentAlive`
//  that stays true through a failure face, and a probe error recorded as success.
//  The TS package owns the machine
//  (`packages/dsh-stream-state/src/load-state.ts`); this file is the SWIFT mirror the
//  shell actually runs, so the phase names, the generation fence and the thresholds
//  have one definition on both sides.
//
//  PURITY. Foundation only, no AppKit/WebKit: the parity gate
//  (`scripts/gates/verify-stream-state-swift-parity.mjs`) compiles this file with a
//  plain `swiftc` on any platform, and the CI macOS leg compiles it with the app.
//
//  SCOPE. The pure machine only. Owning the WKWebView, the timers and the sidecar is
//  the shell's job (design 25); this file decides, it does not execute.
//

import Foundation

/// The phases, in the order a healthy shell visits them.
public enum LoadPhase: String, CaseIterable, Equatable {
  case cold
  case probing
  case loading
  case loaded
  case retrying
  case failurePage
}

public struct LoadEnv: Equatable {
  /// Failed probes before the shell moves to `retrying`.
  public let probeStrikeLimit: Int
  /// Retries before the give-up gate may fire.
  public let retryLimit: Int
  /// Loads this long without content are late (the reported progress SLA).
  public let progressSlaMs: Double

  public init(probeStrikeLimit: Int, retryLimit: Int, progressSlaMs: Double) {
    self.probeStrikeLimit = probeStrikeLimit
    self.retryLimit = retryLimit
    self.progressSlaMs = progressSlaMs
  }
}

public struct LoadState: Equatable {
  public var phase: LoadPhase
  /// Every event belongs to a generation; a stale one is ignored.
  public var generation: Int
  /// Consecutive failed probes in THIS generation (reset by a success).
  public var probeStrikes: Int
  /// A crash recovery is in flight: cleared when content is honestly alive.
  public var recoveringFromCrash: Bool
  /// The one-shot give-up gate: consumed when the failure page is shown.
  public var giveUpSpent: Bool
  /// When the current load armed, for the progress SLA. Nil while not loading.
  public var loadingSinceMs: Double?

  public init(
    phase: LoadPhase, generation: Int, probeStrikes: Int,
    recoveringFromCrash: Bool, giveUpSpent: Bool, loadingSinceMs: Double?
  ) {
    self.phase = phase
    self.generation = generation
    self.probeStrikes = probeStrikes
    self.recoveringFromCrash = recoveringFromCrash
    self.giveUpSpent = giveUpSpent
    self.loadingSinceMs = loadingSinceMs
  }

  public static let initial = LoadState(
    phase: .cold, generation: 0, probeStrikes: 0,
    recoveringFromCrash: false, giveUpSpent: false, loadingSinceMs: nil
  )

  /// The honest answer to "webViewContentAlive": NOT a flag the page sets, but this
  /// predicate. Only `loaded` counts.
  public var contentIsBelievable: Bool { phase == .loaded }

  /// Whether the load has outlived its progress SLA. A predicate: the retry schedule
  /// is what acts on it.
  public func isLate(now: Double, env: LoadEnv) -> Bool {
    guard let since = loadingSinceMs else { return false }
    return now - since > env.progressSlaMs
  }
}

public enum LoadEffect: Equatable {
  case scheduleRecovery(generation: Int)
  case showFailurePage
  case log(name: String, detail: String)
}

public enum LoadEvent: Equatable {
  case generationStarted(generation: Int, at: Double)
  case loadStarted(generation: Int, at: Double)
  case contentAlive(generation: Int, at: Double)
  case probeSucceeded(generation: Int, at: Double)
  case probeFailed(generation: Int, at: Double)
  case recoveryScheduled(generation: Int, at: Double)
  case recoveryFailed(generation: Int, at: Double)
  case crashRecovered(generation: Int, at: Double)

  public var generation: Int {
    switch self {
    case let .generationStarted(g, _), let .loadStarted(g, _), let .contentAlive(g, _),
      let .probeSucceeded(g, _), let .probeFailed(g, _), let .recoveryScheduled(g, _),
      let .recoveryFailed(g, _), let .crashRecovered(g, _):
      return g
    }
  }

  public var at: Double {
    switch self {
    case let .generationStarted(_, t), let .loadStarted(_, t), let .contentAlive(_, t),
      let .probeSucceeded(_, t), let .probeFailed(_, t), let .recoveryScheduled(_, t),
      let .recoveryFailed(_, t), let .crashRecovered(_, t):
      return t
    }
  }

  public var isGenerationStarted: Bool {
    if case .generationStarted = self { return true }
    return false
  }
}

public enum LoadStateMachine {
  /// The generation fence lives here, in one place: every event except a NEW
  /// generation is dropped when it belongs to a superseded one, so a
  /// `didStartLoading` latch cannot exist.
  public static func reduce(
    _ state: LoadState, _ event: LoadEvent, _ env: LoadEnv
  ) -> (state: LoadState, effects: [LoadEffect]) {
    if !event.isGenerationStarted && event.generation != state.generation {
      return (state, [])
    }

    switch event {
    case let .generationStarted(generation, _):
      if generation <= state.generation { return (state, []) }
      var next = state
      next.phase = .cold
      next.generation = generation
      next.probeStrikes = 0
      next.giveUpSpent = false
      next.loadingSinceMs = nil
      return (next, [.log(name: "load-generation", detail: String(generation))])

    case let .loadStarted(_, at):
      var next = state
      next.phase = .loading
      next.loadingSinceMs = at
      return (next, [])

    case .contentAlive:
      var next = state
      next.phase = .loaded
      next.probeStrikes = 0
      next.recoveringFromCrash = false
      next.loadingSinceMs = nil
      return (next, [])

    case .probeSucceeded:
      var next = state
      next.phase = .loaded
      next.probeStrikes = 0
      return (next, [])

    case let .probeFailed(_, at):
      // A failed probe is a STRIKE. It never marks the shell loaded.
      let strikes = state.probeStrikes + 1
      var next = state
      next.probeStrikes = strikes
      if strikes >= env.probeStrikeLimit {
        next.phase = .retrying
        if next.loadingSinceMs == nil { next.loadingSinceMs = at }
        return (next, [
          .log(name: "probe-failed", detail: String(strikes)),
          .scheduleRecovery(generation: state.generation),
        ])
      }
      if state.phase == .loaded { next.phase = .probing }
      return (next, [.log(name: "probe-failed", detail: String(strikes))])

    case let .recoveryScheduled(_, at):
      var next = state
      next.phase = .retrying
      next.recoveringFromCrash = true
      if next.loadingSinceMs == nil { next.loadingSinceMs = at }
      return (next, [])

    case .recoveryFailed:
      // The give-up gate is one-shot per generation.
      if state.giveUpSpent {
        return (state, [.log(name: "recovery-failed-again", detail: String(event.generation))])
      }
      var next = state
      next.phase = .failurePage
      next.giveUpSpent = true
      next.recoveringFromCrash = false
      return (next, [.showFailurePage])

    case .crashRecovered:
      var next = state
      next.recoveringFromCrash = false
      next.phase = .probing
      next.probeStrikes = 0
      return (next, [])
    }
  }
}

File: packages/dsh-stream-state/src/load-state.ts
/**
 * The shell's explicit load state machine.
 *
 * One state machine with an explicit `generation` replaces three scattered booleans
 * whose defects only show on a real machine: a `didStartLoading` latch with no reset
 * point; `webViewContentAlive` true on a failure face; and a probe error recorded as
 * success. Every event carries its generation, and an event from a superseded
 * generation is DROPPED rather than applied - a new generation starts at `cold`.
 * Content is only believable in `loaded`, and a failed probe is a strike, never a
 * success.
 *
 * PURITY: imports NOTHING; the Swift mirror reads `tables.json` and is locked to the
 * same phase names and thresholds.
 */

/** The phases, in the order a healthy shell visits them. `retrying` and
 *  `failurePage` are the two recovery faces. */
export type LoadPhase = 'cold' | 'probing' | 'loading' | 'loaded' | 'retrying' | 'failurePage'

export interface LoadState {
  readonly phase: LoadPhase
  /** Every event belongs to a generation; a stale one is ignored. */
  readonly generation: number
  /** Consecutive failed probes in THIS generation (reset by a success). */
  readonly probeStrikes: number
  /** A crash recovery is in flight: cleared when content is honestly alive. */
  readonly recoveringFromCrash: boolean
  /** The one-shot give-up gate: consumed when the failure page is shown. */
  readonly giveUpSpent: boolean
  /** When the current load armed, for the progress SLA. Null while not loading. */
  readonly loadingSinceMs: number | null
}

export interface LoadEnv {
  /** Failed probes before the shell moves to `retrying`. */
  readonly probeStrikeLimit: number
  /** Retries before the give-up gate may fire. */
  readonly retryLimit: number
  /** Loads this long without content are late (the SLA the shell reports). */
  readonly progressSlaMs: number
}

export type LoadEffect =
  | { readonly e: 'scheduleRecovery'; readonly generation: number }
  | { readonly e: 'showFailurePage' }
  | { readonly e: 'log'; readonly name: string; readonly detail: string }

export type LoadEvent =
  /** A new sidecar/web generation exists: the shell restarts its bookkeeping. */
  | { readonly kind: 'generationStarted'; readonly generation: number; readonly at: number }
  /** WebKit began a load for this generation (`didStartLoading`). */
  | { readonly kind: 'loadStarted'; readonly generation: number; readonly at: number }
  /** The page reports it is alive AND has content. Only believed in `loaded`. */
  | { readonly kind: 'contentAlive'; readonly generation: number; readonly at: number }
  | { readonly kind: 'probeSucceeded'; readonly generation: number; readonly at: number }
  /** A probe FAILED: a strike, never a success. */
  | { readonly kind: 'probeFailed'; readonly generation: number; readonly at: number }
  | { readonly kind: 'recoveryScheduled'; readonly generation: number; readonly at: number }
  | { readonly kind: 'recoveryFailed'; readonly generation: number; readonly at: number }
  | { readonly kind: 'crashRecovered'; readonly generation: number; readonly at: number }

export function initialLoadState(): LoadState {
  return {
    phase: 'cold',
    generation: 0,
    probeStrikes: 0,
    recoveringFromCrash: false,
    giveUpSpent: false,
    loadingSinceMs: null,
  }
}

/** Whether the shell may present real content; NOT a boolean the page sets. */
export function contentIsBelievable(state: LoadState): boolean {
  return state.phase === 'loaded'
}

/** Whether the load has outlived its progress SLA (the shell reports this; the retry
 *  schedule acts on it). LATE IS A PHASE PREDICATE: `loadingSinceMs` is an arming stamp,
 *  so only an ACTIVE load can be late, and a non-finite or rolled-back clock can never
 *  make it late. */
export function loadIsLate(state: LoadState, now: number, env: LoadEnv): boolean {
  if (state.phase !== 'loading' || state.loadingSinceMs === null) return false
  const elapsed = now - state.loadingSinceMs
  return Number.isFinite(elapsed) && elapsed > env.progressSlaMs
}

export function reduceLoadState(
  state: LoadState,
  event: LoadEvent,
  env: LoadEnv,
): { readonly state: LoadState; readonly effects: readonly LoadEffect[] } {
  // The generation fence: everything except a NEW generation is dropped when superseded.
  if (event.kind !== 'generationStarted' && event.generation !== state.generation) {
    return { state, effects: [] }
  }

  switch (event.kind) {
    case 'generationStarted':
      // A restarted sidecar invalidates every fact the old generation carried.
      if (event.generation <= state.generation) return { state, effects: [] }
      return {
        state: {
          ...state,
          phase: 'cold',
          generation: event.generation,
          probeStrikes: 0,
          giveUpSpent: false,
          loadingSinceMs: null,
        },
        effects: [{ e: 'log', name: 'load-generation', detail: String(event.generation) }],
      }

    case 'loadStarted':
      // Arming is per-generation, so a load after a failure is believed again.
      return {
        state: { ...state, phase: 'loading', loadingSinceMs: event.at },
        effects: [],
      }

    case 'contentAlive':
      // Content is only believable in `loaded`, and this event is what puts it there.
      return {
        state: {
          ...state,
          phase: 'loaded',
          probeStrikes: 0,
          recoveringFromCrash: false,
          loadingSinceMs: null,
        },
        effects: [],
      }

    case 'probeSucceeded':
      return { state: { ...state, phase: 'loaded', probeStrikes: 0, loadingSinceMs: null }, effects: [] }

    case 'probeFailed': {
      // A failed probe is a STRIKE. It never marks the shell loaded.
      const probeStrikes = state.probeStrikes + 1
      if (probeStrikes >= env.probeStrikeLimit) {
        return {
          state: { ...state, probeStrikes, phase: 'retrying', loadingSinceMs: state.loadingSinceMs ?? event.at },
          effects: [
            { e: 'log', name: 'probe-failed', detail: String(probeStrikes) },
            { e: 'scheduleRecovery', generation: state.generation },
          ],
        }
      }
      return {
        state: { ...state, probeStrikes, phase: state.phase === 'loaded' ? 'probing' : state.phase },
        effects: [{ e: 'log', name: 'probe-failed', detail: String(probeStrikes) }],
      }
    }

    case 'recoveryScheduled':
      return {
        state: { ...state, phase: 'retrying', recoveringFromCrash: true, loadingSinceMs: state.loadingSinceMs ?? event.at },
        effects: [],
      }

    case 'recoveryFailed': {
      // The give-up gate is one-shot: the failure page shows at most once per generation.
      if (state.giveUpSpent) {
        return { state, effects: [{ e: 'log', name: 'recovery-failed-again', detail: String(event.generation) }] }
      }
      return {
        state: { ...state, phase: 'failurePage', giveUpSpent: true, recoveringFromCrash: false, loadingSinceMs: null },
        effects: [{ e: 'showFailurePage' }],
      }
    }

    case 'crashRecovered':
      return {
        state: { ...state, recoveringFromCrash: false, phase: 'probing', probeStrikes: 0, loadingSinceMs: null },
        effects: [],
      }

    default:
      return { state, effects: [] }
  }
}
